python/bindings/include/openravepy/openravepy_int.h

```diff
@@ -228,7 +231,50 @@ class OPENRAVEPY_API AutoPyArrayObjectDereferencer
     PyArrayObject* _pyarrobj;
 };
 
-typedef OPENRAVE_SHARED_PTR<PythonThreadSaver> PythonThreadSaverPtr;
+class OPENRAVEPY_API PyArrayCapsule
+{
+public:
+    // \brief the only entrypoint that allows a null pointer. Expects reset to follow
+    PyArrayCapsule()
+      : _ptr(nullptr)
+    {
+    }
+
+    // \param ptr a not-null pointer
+    PyArrayCapsule(PyArrayObject* ptr)
+    {
+        reset(ptr);  // Avoid initializer, and use a single entry-point
+    }
+
+    // \param ptr a not-null pointer
+    void reset(PyArrayObject* ptr)
+    {
+        if (!ptr) {
+            throw OPENRAVE_EXCEPTION_FORMAT0(_("Invalid Numpy-array pointer. Failed to get contiguous array?"), ORE_InvalidArguments);
+        }
+        _ptr = ptr;
+    }
+
+    PyArrayObject* get() const
+    {
+        return _ptr;
+    }
+
+    operator PyArrayObject*() const
+    {
+        return _ptr;
+    }
+
+    ~PyArrayCapsule() {
+        if (!_ptr) {
+            return;
+        }
+        Py_DECREF(_ptr);
+    }
+
+private:
+    PyArrayObject* _ptr;
+};
```

Comment:
I think this is more or less equivalent to:

```c++
class PyArrayCapsule : public boost::shared_ptr<PyArrayObject>
{
public:
    PyArrayCapsule(PyArrayObject* ptr) : boost::shared_ptr<PyArrayObject>(ptr, PyArrayCapsule::Deleter()) {}
    PyArrayCapsule() : PyArrayCapsule(nullptr) {}

private:
    class Deleter
    {
    public:
        void operator()(PyArrayObject *ptr) {
            if ( ptr != nullptr ) {
                Py_DECREF(p);
            }
        }
    };
};
```

I would recommend against writing your own shared pointer because it is error-prone. For instance, in your case copy constructor is not deleted and will cause double free.

Reply:
@ziyan I don't think the purpose is to use reference counting here as much as just trigger the destructor, which the old implementation should be fine except for the fact that it throws an exception whenever the pointer is null.
what advantage does boost::shared_ptr have besides reduced code?

Reply:
It prevents this misuse:

```c++
PyArrayCapsule SomeFunction() {
    PyArrayCapsule cap(...);
    return cap
}
```

Or

```c++
PyArrayCapsule a(...);
PyArrayCapsule b = a;
```

The current implementation, the above 2 patterns will result in `Py_DECREF` being called twice in each case.

Reply:
@ziyan hmm... worrying about usages like that means even the current implementation has this problem, but i highly doubt anyone will do that. nevertheless, i made your point.
would the following one-liner work?

```c++
boost::shared_ptr<PyArrayObject> parray(PyArray_GETCONTIGUOUS(reinterpret_cast<PyArrayObject*>(obj.ptr())), Py_DECREF);
```

Reply:
Py_DECREF is not a class, so you have to have a deleter class

Reply:
`Py_DECREF` just needs to be a function for this to work..

Reply:
looks like it is a macro ;0/

```c++
#define Py_DECREF(op)                                   \
    do {                                                \
        if (_Py_DEC_REFTOTAL  _Py_REF_DEBUG_COMMA       \
        --((PyObject*)(op))->ob_refcnt != 0)            \
            _Py_CHECK_REFCNT(op)                        \
        else                                            \
        _Py_Dealloc((PyObject *)(op));                  \
    } while (0)

```


however, there's also a function
```
PyAPI_FUNC(void) Py_DecRef(PyObject *);
```

perhaps that can be used instead..?

Reply:
Oh, I didn't know that. Then yes, you really just need to use it like this:

```c++
boost::shared_ptr<PyArrayObject> a(..., Py_DECREF);
```

For convienence, consider:

```c++
class PyArrayCapsule : public boost::shared_ptr<PyArrayObject>
{
public:
    PyArrayCapsule(PyArrayObject* ptr) : boost::shared_ptr<PyArrayObject>(ptr, Py_DECREF) {}
    PyArrayCapsule() : PyArrayCapsule(nullptr) {}
};
```

Reply:
Perhaps this?

```c++
template<class T>
class PyCapsule : public boost::shared_ptr<T>
{
public:
    PyCapsule(T* ptr) : boost::shared_ptr<T>(ptr, PyCapsule<T>::_Delete) {}
    PyCapsule() : PyCapsule(nullptr) {}

private:
    static void _Delete(T *ptr) { Py_DECREF(ptr); }
};

typedef PyCapsule<PyArrayObject> PyArrayCapsule;
```

Reply:
@ziyan @rdiankov I adapted the proposal according to your valuable comments. Many thanks~
- prevented the copy constructor to prevent more than a single decrement
- made the it a template class to support more use cases

However, I think using the smart pointer is an overkill for this small task, especially with its atomic and thread-safety stuff.
 More importantly, this class is not going to be used in other projects, in which explicitly raising on cases of nullptr being passed is really needed, mostly to catch bugs easily, thus I wonder if you can be fine with the current version?